onnxruntime/contrib_ops/cpu/qlinear_concat.cc

```diff
@@ -0,0 +1,139 @@
+
+#include "qlinear_concat.h"
+#include "contrib_ops/cpu/qlinear_lookup_table.h"
+#include "core/providers/common.h"
+#include "core/mlas/inc/mlas.h"
+#include "core/platform/threadpool.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+QLinearConcat::QLinearConcat(const OpKernelInfo& info) : OpKernel(info), ConcatBase(info) {
+  size_t input_def_count = info.node().InputDefs().size();
+  ORT_ENFORCE(input_def_count >= 8 && (input_def_count - 2) % 3 == 0,
+              "At lease two inputs are needed, and each input must be (tensor, scale, zero_point) tuple!");
+
+  size_t input_count = (input_def_count - 2) / 3;
+  fixed_lookup_tables_.resize(input_count);
+
+  const Tensor* tensor_y_scale = nullptr;
+  const Tensor* tensor_y_zero_point = nullptr;
+  bool get_y_scale = info.TryGetConstantInput(0, &tensor_y_scale);
+  bool get_y_zero_point = info.TryGetConstantInput(1, &tensor_y_zero_point);
+  if (!get_y_scale || !get_y_zero_point) {
+    // Can not build const lookup table
+    return;
+  }
+
+  // Initialize lookup table given constant input/output scales and zero points
+  bool is_signed_int8 = tensor_y_zero_point->IsDataType<int8_t>();
+  const auto identity_float = [](float v) -> float { return v; };
+  for (size_t def_index = 2; def_index < input_def_count; def_index += 3) {
+    const Tensor* tensor_x_scale = nullptr;
+    const Tensor* tensor_x_zero_point = nullptr;
+    bool get_x_scale = info.TryGetConstantInput(static_cast<int>(def_index) + 1, &tensor_x_scale);
+    bool get_x_zero_point = !info.TryGetConstantInput(static_cast<int>(def_index) + 2, &tensor_x_zero_point);
+    if (!get_x_scale || !get_x_zero_point) {
+      continue;  // try to optimize next one
+    }
+    ORT_ENFORCE(tensor_x_scale->IsDataType<float>(), "Input scale is not float for input def @", def_index + 1);
+    ORT_ENFORCE(tensor_x_zero_point->GetElementType() == tensor_y_zero_point->GetElementType(),
+                "Wrong input type encountered for zero point input def @", def_index + 2);
+
+    size_t input_idx = (def_index - 2) / 3;
+    fixed_lookup_tables_[input_idx].resize(256);
+    if (is_signed_int8) {
+      QlinearBuildLookupTable<int8_t>(
+          fixed_lookup_tables_[input_idx].data(), tensor_x_scale, tensor_x_zero_point,
+          tensor_y_scale, tensor_y_zero_point, identity_float);
+    } else {
+      QlinearBuildLookupTable<uint8_t>(
+          fixed_lookup_tables_[input_idx].data(), tensor_x_scale, tensor_x_zero_point,
+          tensor_y_scale, tensor_y_zero_point, identity_float);
+    }
+  }
+}
+
+Status QLinearConcat::Compute(OpKernelContext* ctx) const {
+  const Tensor* tensor_y_scale = ctx->Input<Tensor>(0);
+  const Tensor* tensor_y_zero_point = ctx->Input<Tensor>(1);
+  bool is_signed_int8 = tensor_y_zero_point->IsDataType<int8_t>();
+  const auto identity_float = [](float v) -> float { return v; };
+
+  // Number of input tensors to concatenate (tupled)
+  auto input_count_x3 = Node().InputArgCount()[2];
```

Comment:
nit: those info can be determined in construct time

Reply:
Agree. Yet I feel it is worthless to add such member in the class definition. Let's keep get it dynamically.